src/Microsoft.AspNetCore.Hosting/Internal/RequestDiagnosticsMiddleware.cs

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    public class RequestDiagnosticsMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private HostingApplicationDiagnostics _diagnostics;
+
+        public RequestDiagnosticsMiddleware(RequestDelegate next, DiagnosticListener diagnosticListener, ILogger logger)
+        {
+            _next = next;
+            _diagnostics = new HostingApplicationDiagnostics(logger, diagnosticListener);
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            var context = new HostingApplication.Context();
+            _diagnostics.BeginRequest(httpContext, ref context);
+
+            try
+            {
+                await _next.Invoke(httpContext);
+
+                _diagnostics.RequestEnd(httpContext, null, context);
+            }
+            catch (Exception exception)
+            {
+                _diagnostics.RequestEnd(httpContext, exception, context);
```

Comment:
Moving this to middleware loses quite a bit of fidelity for the error case, and a little fidelity even for normal requests.

When an exception is thrown before the first write:
2.1: the server changes the status code to 500, clears the headers, send the response, and then logs RequestEnd with a 500 and no headers.
New: RequestEnd is logged with a 200 status code and standard response headers, then the server changes the status code to 500, clears the headers, send the response.

This is too breaking, the logs will be much less accurate.

This also makes the timing less accurate as the server used to do more work before logging RequestEnd (flushes headers, send the final response chunk, etc.)

Reply:
Would running request end in OnCompleted help here? We don’t flow the exception there but it could be captured in the middleware.

Reply:
OnCompleted would be better timing wise, that's [immediately before DisposeContext](https://github.com/aspnet/KestrelHttpServer/blob/edc19354751dd43dc48e71cdfe192147e40122c8/src/Kestrel.Core/Internal/Http/HttpProtocol.cs#L602-L607). You will still miss out on logging a few exceptions from places like [OnStarting](https://github.com/aspnet/KestrelHttpServer/blob/edc19354751dd43dc48e71cdfe192147e40122c8/src/Kestrel.Core/Internal/Http/HttpProtocol.cs#L700-L716) and [ContentLength](https://github.com/aspnet/KestrelHttpServer/blob/edc19354751dd43dc48e71cdfe192147e40122c8/src/Kestrel.Core/Internal/Http/HttpProtocol.cs#L894) though.


It would require additional allocations per request to register and pass state though.